Generate serializers for NSArray, NSDictionary, and NSFont

beidson opened the pull request:
#### 6be1434f578ebcb90ea96a1c71bca5348812e60a
<pre>
Generate serializers for NSArray, NSDictionary, and NSFont
<a href="https://bugs.webkit.org/show_bug.cgi?id=264696">https://bugs.webkit.org/show_bug.cgi?id=264696</a>

Reviewed by Alex Christensen.

Now that we have the base CoreIPCNSCFObject wrapper and .serialization.in files for fundamental NS types, we can:
- Turn NSArray into a &quot;Vector of CoreIPCNSCFObjects&quot;
- Turn NSDictionary into a &quot;Vector of key/value pairs where the keys and values are CoreIPCNSCFObjects&quot;
- Directly serialize NSFont/UIFont, since we used an NSDictionary to do so before

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::isSerializableValue):
(IPC::encodeObjectDirectly&lt;NSArray&gt;): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSArray&gt;): Deleted.
(IPC::encodeObjectDirectly&lt;NSDictionary&gt;): Deleted.
(IPC::id&gt;&gt;): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSDictionary&gt;): Deleted.
(IPC::encodeObjectDirectly&lt;WebCore::CocoaFont&gt;): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;WebCore::CocoaFont&gt;): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h.
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCFont.h:
(WebKit::CoreIPCFont::CoreIPCFont):
(WebKit::CoreIPCFont::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCFont.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCFont::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCFont.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::CoreIPCNSCFObject::valueIsAllowed):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
(WebKit::CoreIPCSecureCoding::objectClass):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(operator==):
(TEST):

Canonical link: <a href="https://commits.webkit.org/270624@main">https://commits.webkit.org/270624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f67fcef417335143f034668281b06e238a4cf5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4567 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28056 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26274 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6329 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1995 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23849 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22362 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28636 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29379 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23683 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23692 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27256 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3101 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1309 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4491 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6237 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3558 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3419 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
```diff
@@ -32,24 +32,27 @@
-    }
-
+    CoreIPCArray(NSArray *);
     CoreIPCArray(RetainPtr<NSArray> array)
```
could be const &

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
```diff
@@ -32,24 +32,27 @@
+    {
+    }
+
+    Vector<std::unique_ptr<CoreIPCNSCFObject>> m_array;
```
I don't see any reason to have a unique_ptr here.  Just have a Vector<CoreIPCNSCFObject>

beidson replied:
TBQH it was to fight forward declaration vs actual includes of the header, as CoreIPCNSCFObject.h includes all of the subordinate headers.

I'll try to rework forward declares vs includes to get this simpler.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
```diff
@@ -28,27 +28,35 @@
     friend struct IPC::ArgumentCoder<CoreIPCDictionary, void>;
 
-    IPC::CoreIPCRetainPtr<NSDictionary> m_dictionary;
+    using ValueType = Vector<std::pair<std::unique_ptr<CoreIPCNSCFObject>, std::unique_ptr<CoreIPCNSCFObject>>>;
```
Ditto.  Could use WTF::KeyValuePair, don't see need for unique_ptr.  Use UniqueRef if you have to have another allocation, because there's certainly no need for a null state.

It may also be possible that we only need to support keys that are Strings.

beidson replied:
Today, we absolutely do NSDictionaries with arbitrary keys.
I don't think this patch is the right place to rethink that (going for zero behavior change)

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
```diff
@@ -84,6 +84,20 @@
+    Class objectClass;
+    WTF::switchOn(value, [&](CoreIPCSecureCoding& object) {
+        objectClass = object.objectClass();
+    }, [&](auto& object) {
```
This should be explicitly std::nullptr_t instead of this mysterious auto&

beidson replied:
Note: Unlike the previous switchOn, this one is only trying to capture SecureCoding - It's intended that all other cases fall through to a null Class.

beidson committed:
Generate serializers for NSArray, NSDictionary, and NSFont
https://bugs.webkit.org/show_bug.cgi?id=264696

Reviewed by Alex Christensen.

Now that we have the base CoreIPCNSCFObject wrapper and .serialization.in files for fundamental NS types, we can:
- Turn NSArray into a "Vector of CoreIPCNSCFObjects"
- Turn NSDictionary into a "Vector of key/value pairs where the keys and values are CoreIPCNSCFObjects"
- Directly serialize NSFont/UIFont, since we used an NSDictionary to do so before

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::isSerializableValue):
(IPC::encodeObjectDirectly<NSArray>): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSArray>): Deleted.
(IPC::encodeObjectDirectly<NSDictionary>): Deleted.
(IPC::id>>): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSDictionary>): Deleted.
(IPC::encodeObjectDirectly<WebCore::CocoaFont>): Deleted.
(IPC::decodeObjectDirectlyRequiringAllowedClasses<WebCore::CocoaFont>): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h.
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCFont.h:
(WebKit::CoreIPCFont::CoreIPCFont):
(WebKit::CoreIPCFont::toID): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCFont.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCFont::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCFont.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::CoreIPCNSCFObject::valueIsAllowed):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
(WebKit::CoreIPCSecureCoding::objectClass):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(operator==):
(TEST):

Canonical link: https://commits.webkit.org/270624@main
